design/dpop.md

```diff
@@ -0,0 +1,259 @@
+
+## Implementation Details
+
+DPoP is a cross-cutting concept that would affect multiple Keycloak components, namely core libraries, OAuth/OIDC endpoints, adapters, models and admin UI.
+
+It is suggested that DPoP should be set up on a per-client basis, with the three options available that should be interpreted as follows:
+
+|**Value** | **Keycloak Server** | **Keycloak Adaptors** |
+| --- | --- | --- |
+| Required | OAuth endpoints (where applicable) should require a DPoP header and a DPoP-bound token; should return an error otherwise. | Protected resources should require a DPoP header and a DPoP-bound token, or should return an error otherwise. |
+| Optional | _Token Endpoint_: if DPoP header is present, the returned token should be DPoP-bound and should contain a `cnf` claim with a JWK thumbprint; should be a regular token otherwise  <br/><br/>_UserInfo endpoint_: if the access token is DPoP-bound, a matching DPoP header should be required; no header should be required otherwise. | Protected resources should require a matching DPoP header if the supplied access token is DPoP-bound; should ignore the header otherwise. |
+| Disabled | DPoP headers should be ignored; issued tokens shouldn’t contain JWK thumbprints. | DPoP headers and token JWK thumbprints should be ignored. |
+
+Other DPoP configuration options should include:
+
+| **Option** | **Description** |
+| --- | --- |
+| DPoP Proof Lifetime | Time window in which the respective DPoP proof JWT would be accepted |
+| DPoP Allowed Clock Skew | To accommodate for clock offsets, the server MAY accept DPoP proofs that carry an `iat` time in the reasonably near future (e.g., a few seconds in the future). |
+
+_([8.1. DPoP Proof Replay][4])_
+
+### Endpoints
+
+#### Token Endpoint
+
+With DPoP enabled, token endpoint must understand the DPoP header and perform binding between the keypair presented thereby and the issued token.
+
+As per the spec,
+
+> To request an access token that is bound to a public key using DPoP,
+   the client MUST provide a valid DPoP proof JWT in a "DPoP" header
+   when making an access token request to the authorization server's
+   token endpoint.  This is applicable for all access token requests
+   regardless of grant type (including, for example, the common
+   "authorization\_code" and "refresh\_token" grant types but also
+   extension grants such as the JWT authorization grant \[RFC7523\]).
+
+_([5.5. DPoP Access Token Request][6])_
+
+> :question: As DPoP is primarily meant to protect tokens issued to SPAs, should we also support DPoP for:
+>	
+> * password grant 
+> * client credentials grant (DPoP-enabled service accounts?)
+> * token exchange
+> * extension grants (with the future Grant Type SPI)
+> * UMA permissions grant?
+
```

Comment:
IMO, it might be preferable that this document only treats basic authorization code grant and token refresh. Other grant could be realized by follow-up tasks after this document is merged and corresponding PR is also merged.
The reason why is that if this document treats all grants shown here, the corresponding PR's code volume would become huge, which makes the keycloak development team hard to review it and might takes a lot of time for this PR to be merged.

Reply:
Agreed, will move other grant types to milestones.